astropy/io/fits/fitsrec.py

```diff
@@ -1032,7 +1040,7 @@ def _get_raw_data(self):
         May return ``None`` if no array resembling the "raw data" according to
         the stated criteria can be found.
         """
-        raw_data_bytes = self.nbytes + self._heapsize
+        raw_data_bytes = self._tbsize + self._heapsize
```

Comment:
Main fix is here, with the propagation of heap size and the original array size.